articles/hajimehoshi.re

```diff
@@ -0,0 +1,102 @@
+
+== @<tt>{gomobile}
+
+@<tt>{gomobile}はAndroidおよびiOS向けのアプリケーションもしくはライブラリを出力するソフトウェアです。
+パッケージは@<tt>{golang.org/x/mobile}です。
+
+なぜ@<tt>{gomobile}がGo本体以外に必要になるかと言うと、Go本体ではモバイル向けのアプリケーションやライブラリを作れないからです。
+Go本体では、AndroidとiOS向けに、C互換の共有ライブラリまたは静的ライブラリを出力することができます。
+実際@<tt>{gomobile}はGo本体にコンパイル処理を丸投げしています。
+@<tt>{gomobile}はGoのコードを単にコンパイルするだけではなく、AndroidやiOSアプリケーションとして使いやすい形に変換します。
+たとえばAndroidが使えるライブラリは@<tt>{.aar}形式であり、特定のファイルをzipで固めたものになりますが、
+Go本体はこれを生成してくれません。
+また、たとえばGoの関数をJavaで使う際には、バインディングのためのコードが別途必要になってしまいます。
+@<tt>{gomobile}はそれらをやっているわけです。
+
+@<tt>{gomobile}コマンドは大雑把に2つのモードがあります。
+@<tt>{gomobile build}と@<tt>{gomobile bind}です。
+
+@<tt>{gomobile build}は単独のアプリケーションを出力します。
+Android向けには@<tt>{.apk}ファイル、iOS向けには@<tt>{.ipa}ファイルです。
+アプリケーション全体をすべてGoで書くことが可能です。
+機能は限定されていて、アプリケーションはOpenGLサーフェスを一個だけ持ちます。
+
+一方@<tt>{gomobile bind}はライブラリを出力します。
+Android向けには@<tt>{.aar}ファイル、iOS向けには@<tt>{.framework}ファイルです。
+Goのexportされた関数などは、一定の条件@<fn>{gobind}を満たしたものは、JavaもしくはObjective-C側でも呼び出せるようになります。
+Goで定義したinterfaceをJava側で実装してGoの関数に渡す、といったことも可能です。
+あくまで共有ライブラリなので、これ単独では完成したアプリケーションになりません。
+ユーザーはGo以外の言語を書く必要があります。
+
+//footnote[gobind][https://godoc.org/golang.org/x/mobile/cmd/gobind]
+
+@<tt>{gomobile build}は簡単に完成されたアプリケーションを作れる一方、柔軟性に欠けます。
+@<tt>{gomobile build}で作られたアプリケーションはOpenGLサーフェス一個だけしかもっておらず、
+他のコンポーネントを入れることができません。またストアに必要な署名などのファイルを挿入するのは、
+不可能ではないにしろ困難です。
+@<tt>{gomobile build}だけで、ストアに載せる実用アプリケーションを作ることは現実的ではありません@<fn>{tenntenn}。
+実際、我々が開発しているゲームでは@<tt>{gomobile bind}を活用しています。
+
+//footnote[tenntenn][実際に@<tt>{gomobile build}で作られたアプリケーションをストアに載せた例もあります: @<href>{https://www.slideshare.net/takuyaueda967/go-mobileandroid}]
+
+== @<tt>{gomobile}と@<tt>{gobind}
+
+@<tt>{golang.org/x/mobile/cmd}配下には2つコマンドが存在します。@<tt>{gomobile}と@<tt>{gobind}です。@<tt>{gomobile}は内部的に@<tt>{gobind}を呼び出しています。
+
+普段ユーザーが使うコマンドは@<tt>{gomobile}です。@<tt>{gobind}を直接使用することも可能です。
+
+@<tt>{gobind}コマンドはバインディング用のソースコードを生成します。
+指定されたパッケージで、一定条件を満たした関すなどはJavaやObjective-Cに露出しますが、それの橋渡しとなる部分を生成します。
+またリバースバインディングに関するソースコードも生成します。
+
+一方@<tt>{gomobile}は@<tt>{gobind}で生成されたソースコードをGo本体でコンパイルし、パッケージ化します。
+ソースコードをコンパイルするのはGo本体であって、@<tt>{gomobile}はあくまでGoを呼び出しているだけです。
+
+== @<tt>{gomobile}の現状
+
+@<tt>{gomobile}はElias Naur氏@<fn>{elias}がメンテナーとして活躍されていましたが、
+最近メンテナーを辞めてしまい、誰もアクティブにメンテナンスしていないという状況でした。
+そこで僭越ながら私がメンテナーとして立候補しました。
+今は最重要課題であると思われるModule対応@<fn>{module}をがんばっています。
+
+//footnote[elias][@<href>{https://github.com/eliasnaur}]
+//footnote[module][@<href>{https://github.com/golang/go/issues/27234}]
+
+ちなみにElias氏はGio@<fn>{gio}というGUIライブラリを作成しています。
+GioはAndroidやiOS向けアプリケーションを生成できます。
+しかしながら@<tt>{gomobile}は一切使われていません。
+さきほど述べたとおり@<tt>{gomobile}を一切使わずとも自力でパッケージを作ることが可能です。
+実際自力でパッケージを作ることができ、かつバインディング部分も自力で代用できるのであれば
+@<tt>{gomobile}を使う必要がないわけです。
+興味のある方はGioのソースコードを読んでみるとよいでしょう。
+
+//footnote[gio][@<href>{https://gioui.org}]
+
+== @<tt>{gobind}が何をしているか
+
+TODO: 書く。
+
+== @<tt>{gomobile}が何をしているか
+
+TODO: 書く。
+
+== 応用例: @<tt>{ebitenmobile}
+
+TODO: 書く?
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "?". (ja-technical-writing/no-exclamation-question-mark)